Add suggestions for relating types in enum property search

falken10vdl opened the pull request:
Here is a proposal for https://github.com/IfcOpenShell/IfcOpenShell/issues/5166
Cheers!

falken10vdl committed:
Add suggestions for relating types in enum property search

falken10vdl committed:
Include 'relating_type_id' for suggestions

Andrej730 commented:
@falken10vdl 

1. `if element and element.is_a().endswith("Type"):` - what this for?

2. `getattr(element, "Name", "") or "Unnamed"` - are there cases when Name is missing?

3. `value != "NOTDEFINED"` - why `NOTDEFINED` is special?

4. `next((getattr(element, "PredefinedType", None) for _ in [1]), "NOTDEFINED")` - what's this for?

5. Can you please add it as an option to `prop_with_search`, rather than hardcoding it to `("relating_type", "relating_type_id")`?

falken10vdl committed:
Adding changes as per core developers feedback

falken10vdl commented:
@Andrej730 My comments thanks!

>     if element and element.is_a().endswith("Type"): - what this for?
To make sure that the element are "types" and not instances. Foir example IfcWallType but not IfcWall The request was to add suggestions for ifcElmentTypes


>     getattr(element, "Name", "") or "Unnamed" - are there cases when Name is missing?
It is just defensive coding but I think there could be cases where the ifc schema has Name as optional or simply the ifc file is not well formated. This is just to ensure that the UI gets a readable friendly suggestion but I agree in most of the cases is not needed. Please advice if it is better to remove


>     value != "NOTDEFINED" - why NOTDEFINED is special?
Well because adding "NOTDEFINED" to the suggestions seems not very informative. Example:
PredefinedType = "NOTDEFINED"
Description = "Exterior Wall"
Only "Exterior Wall" would be included in the suggestion
This can be removed and the "NOTDEFINED" will appear as a suggestion if you think is more coherent



>     next((getattr(element, "PredefinedType", None) for _ in [1]), "NOTDEFINED") - what's this for?
It is the same as getattr(element, "PredefinedType", None) or "NOTDEFINED" just a way to handle the case where PredefinedType might be None and provide "NOTDEFINED" as a fallback
Changed.


>     Can you please add it as an option to prop_with_search, rather than hardcoding it to ("relating_type", "relating_type_id")?
In a smiliar way to how should_click_ok is done? Please take a look


Why only for >10?
`        if len(get_enum_items(data, prop_name, original_operator_path=original_operator_path)) > 10:`

With suggestions I believe it is usefull to get the search with suggestions even for fewer number of elements since the name could be cryptic

Cheers!

Andrej730 commented:
@falken10vdl 

> To make sure that the element are "types" and not instances. Foir example IfcWallType but not IfcWall The request was to add suggestions for ifcElmentTypes

But how a non-type can end up in a relating type property?

> It is just defensive coding but I think there could be cases where the ifc schema has Name as optional or simply the ifc file is not well formated. This is just to ensure that the UI gets a readable friendly suggestion but I agree in most of the cases is not needed. Please advice if it is better to remove

`or "Unnamed"` makes sense, but `getattr` is misleading - it suggests that there are cases when .Name attribute will be missing completely.

> Well because adding "NOTDEFINED" to the suggestions seems not very informative. Example:

I see, basically it's the same as `None` then. Okay, let's skip it.

> It is the same as getattr(element, "PredefinedType", None) or "NOTDEFINED" just a way to handle the case where PredefinedType might be None and provide "NOTDEFINED" as a fallback

Makes sense, it's just that `next` and `for _ in [1]` construction was kind of odd.

> In a smiliar way to how should_click_ok is done? Please take a look

Yes, the same way, argument + prop.

> Why only for >10?

I agree it can be useful to have less than 10 in this case, as suggestions have additional information besides just name. Let's expose it as `prop_with_search` argument, so it would show up if there's more than 0 items.

falken10vdl committed:
Add changes as per core developer's feedback

falken10vdl commented:
> But how a non-type can end up in a relating type property

I guess it could be the case for a bad formed ifc file. Removed to make it simpler

> or "Unnamed" makes sense, but getattr is misleading - it suggests that there are cases when .Name attribute will be missing completely.

ok. Changed,

> I see, basically it's the same as None then. Okay, let's skip it.

Removed all the NOTDEFINED handling. It will appear the predifiendtype NOTDEFINED in the suggestions as well.

> I agree it can be useful to have less than 10 in this case, as suggestions have additional information besides just name. Let's expose it as prop_with_search argument, so it would show up if there's more than 0 items.

added search_threshold argument (default to 0)

Andrej730 commented:
> added search_threshold argument (default to 0)

It should default to the previous value, `10`, to prevent altering existing UI.

falken10vdl committed:
set default search threshold in prop_with_search function to previous value of 10

falken10vdl commented:
> It should default to the previous value, 10, to prevent altering existing UI.

changed to 10

Cheers!
